syntax = "proto3";

package bloock;

option go_package = "github.com/bloock/bloock-sdk-go/v2/internal/bridge/proto";
option java_package = "com.bloock.sdk.bridge.proto";

message Credential {
    repeated string context = 1;
    string id = 2;
    repeated string type = 3;
    string issuance_date = 4;
    string expiration = 5;
    string credential_subject = 6;
    CredentialStatus credential_status = 7;
    string issuer = 8;
    CredentialSchema credential_schema = 9;
    CredentialProof proof = 10;
}

message CredentialProof {
    string signature_proof = 1;
    optional string sparse_mt_proof = 2;
}

message CredentialStatus {
    string id = 1;
    int64 revocation_nonce = 2;
    string type = 3;
}

message CredentialSchema {
    string id = 1;
    string type = 2;
}

message StringAttribute {
    string id = 1;
    string value = 2;
}

message IntegerAttribute {
    string id = 1;
    int64 value = 2;
}

message DecimalAttribute {
    string id = 1;
    double value = 2;
}

message BooleanAttribute {
    string id = 1;
    bool value = 2;
}

message DateAttribute {
    string id = 1;
    string value = 2;
}

message DateTimeAttribute {
    string id = 1;
    string value = 2;
}

message StringAttributeDefinition {
    string display_name = 1;
    string id = 2;
    string description = 3;
    bool required = 4;
}

message IntegerAttributeDefinition {
    string display_name = 1;
    string id = 2;
    string description = 3;
    bool required = 4;
}

message DecimalAttributeDefinition {
    string display_name = 1;
    string id = 2;
    string description = 3;
    bool required = 4;
}

message BooleanAttributeDefinition {
    string display_name = 1;
    string id = 2;
    string description = 3;
    bool required = 4;
}

message DateAttributeDefinition{
    string display_name = 1;
    string id = 2;
    string description = 3;
    bool required = 4;
}

message DateTimeAttributeDefinition {
    string display_name = 1;
    string id = 2;
    string description = 3;
    bool required = 4;
}

message StringEnumAttributeDefinition {
    string display_name = 1;
    string id = 2;
    string description = 3;
    bool required = 4;
    repeated string enum = 5;
}

message IntegerEnumAttributeDefinition {
    string display_name = 1;
    string id = 2;
    string description = 3;
    bool required = 4;
    repeated int64 enum = 5;
}

message DecimalEnumAttributeDefinition {
    string display_name = 1;
    string id = 2;
    string description = 3;
    bool required = 4;
    repeated double enum = 5;
}

message CredentialReceipt {
    Credential credential = 1;
    string credential_id = 2;
    string credential_type = 3;
}

message IssuerStateReceipt {
    string tx_hash = 1;
}

message Schema {
    string cid = 1;
    string cid_json_ld = 2;
    string schema_type = 3;
    string json = 4;
}

message CredentialRevocation {
    bool success = 1;
}

message VerificationReceipt {
    string verification_request = 1;
    int64 session_id = 2;
}

message DidType {
    Method method = 1;
    Blockchain blockchain = 2;
    NetworkId network_id = 3;
}

enum Method {
    IDEN3 = 0;
    POLYGON_ID = 1;
}

enum Blockchain {
    ETHEREUM = 0;
    POLYGON = 1;
    UNKNOWN_CHAIN = 2;
    NO_CHAIN = 3;
}

enum NetworkId {
    MAIN = 0;
    MUMBAI = 1;
    GOERLI = 2;
    UNKNOWN_NETWORK = 3;
    NO_NETWORK = 4;
}

enum PublishInterval {
    INTERVAL_5 = 0;
    INTERVAL_15 = 1;
    INTERVAL_60 = 2;
}

message SignatureJWS {
    string signature = 1;
    string protected = 2;
    SignatureHeaderJWS header = 3;
    string message_hash = 4;
}

message SignatureHeaderJWS {
    string alg = 1;
    string kid = 2;
    optional string subject = 3;
    optional string hash_alg = 4;
}